EIPS/eip-custom_errors_for_erc_tokens.md

```diff
@@ -0,0 +1,504 @@
+---
+eip: [assignedEIPNumber]
+title: Custom errors for ERC tokens
+description: List of Solidity custom errors covering standard token implementations (EIP-20, EIP-721 and EIP-1155)
+author: Ernesto García (@ernestognw), Francisco Giordano (@frangio), Hadrien Croubois (@Amxx)
+discussions-to: <URL>
+status: Draft
+type: Standards Track
+category: ERC
+requires: 20, 721, 1155
+created: <date created on, in ISO 8601 (yyyy-mm-dd) format>
+---
+
+## Abstract
+
+The following specification allows for the use of a standard list of Solidity [custom errors](https://blog.soliditylang.org/2021/04/21/custom-errors/) to be used within [EIP-20](./eip-20.md), [EIP-721](./eip-721.md) and [EIP-1155](./eip-1155.md) tokens.
+
+Ethereum applications (DApps) and wallets have historically used [revert statements](https://docs.soliditylang.org/en/v0.8.16/control-structures.html#revert-statement) to show relevant failure data to the users, however, this EIP provides a standard list of errors designed to give at least the same relevant information, but in a structured and expected way.
+
+## Motivation
+
+Since the introduction of Solidity custom errors in [v0.8.4](https://github.com/ethereum/solidity/releases/tag/v0.8.4), these have provided a way to show failures in a more expresive way with dynamic arguments, while reducing deployment costs.
+
+At the moment of the release of custom errors, the standard tokens ([EIP-20](./eip-20.md), [EIP-721](./eip-721.md), [EIP-1155](./eip-1155.md)) were already in finalized state, so no error specification was included.
+
+An error specification will allow users to expect more consistent error messages across applications or testing environments, while exposing pertinent arguments and overall reducing the need of writing expensive revert strings in the deployment bytecode.
+
+The list of token standard errors is defined in the specification below.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.
+
+The provided interfaces with errors SHOULD conform with the [error grammar rules](#error-grammar-rules) described in the [rationale](#rationale) section.
+
+It's important to note that errors MAY be thrown in each scenario, but that's up to the implementers, unless a revert reason is specified in their corresponding EIPs.
+
+This EIP aims to provide a clear standard only in case an error will be thrown.
+
+### [EIP-20](./eip-20.md)
+
+#### `ERC20InvalidSender(address)`
```

Comment:
We should include parameter names here:

```suggestion
#### `ERC20InvalidSender(address sender)`
```

Reply:
I remove them because some of them became very large, (eg. `ERC1155InsufficientBalance`).
Not sure if it's the best reading experience, also the parameters are outlined below each error.

What do you think?

Reply:
I really think parameters names are necessary in the title. Haven't seen how large it becomes for insufficient balance but if it is really that large I'd consider giving that function special treatmen.

Reply:
No problem, I added it. The title is large but I don't think that's a problem.